presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java

```diff
@@ -405,6 +435,66 @@ public int getIndex()
         INTERIM,
     }
 
+    public static class DereferenceAdaptation
+    {
+        private final List<HiveColumnHandle> columns;
+        private final Map<Integer, Integer> baseColumnToChannel;
+
+        public DereferenceAdaptation(List<HiveColumnHandle> columns, Map<Integer, Integer> baseColumnToChannel)
+        {
+            this.columns = columns;
+            this.baseColumnToChannel = baseColumnToChannel;
+        }
+
+        public Page adaptPage(Page input)
+        {
+            Block[] blocks = new Block[columns.size()];
+
+            for (int i = 0; i < columns.size(); i++) {
+                HiveColumnHandle column = columns.get(i);
+                int channel = baseColumnToChannel.get(column.getBaseHiveColumnIndex());
+                blocks[i] = createNestedLazyBlock(input.getBlock(channel), column);
+            }
+
+            return new Page(blocks);
+        }
+
+        public static Block createNestedLazyBlock(Block base, HiveColumnHandle column)
+        {
+            List<Integer> dereferences = column.getDereferenceFieldIndices();
+            return new LazyBlock(base.getPositionCount(), new DereferenceBlockLoader(base, dereferences));
+        }
+    }
+
+    public static class DereferenceBlockLoader
+            implements LazyBlockLoader
+    {
+        boolean loaded;
+        private Block parentBlock;
+        private final List<Integer> dereferences;
+
+        DereferenceBlockLoader(Block parentBlock, List<Integer> dereferences)
+        {
+            this.parentBlock = parentBlock;
+            this.dereferences = dereferences;
+        }
+
+        @Override
+        public Block load()
+        {
+            checkState(!loaded, "Already loaded");
+            Block currentBlock = parentBlock;
+            for (int i = 0; i < dereferences.size(); i++) {
+                currentBlock = toColumnarRow(currentBlock).getField(dereferences.get(i));
```

Comment:
@wagnermarkd `toColumnarRow` method uses lazyBlock.getBlock() in case the input block is a lazy block. After changes in #1629 , `LazyBlock::getBlock`, invocation does not load the underlying blocks recursively. (unlike `getLoadedBlock`)   hence the `toColumnarRow` invocation does not load everything. the last `getLoadedBlock` call loads the required dereferenced block. 

The "adapatation" here is about transforming the block to correspond to the required nested type, rather than making it lazy.